In [2]:
# Run this in a Colab cell to set up SQLite from your Excel file
import pandas as pd
import sqlite3

df = pd.read_csv('/content/cleaned_dataset.csv')
conn = sqlite3.connect('orders.db')
df.to_sql('orders', conn, if_exists='replace', index=False)

1200

In [3]:
def run_query(sql):
    result = pd.read_sql_query(sql, conn)
    print(result.to_string(index=False))
    return result

***SELECT Queries***

In [6]:
run_query("SELECT * FROM orders LIMIT 5")

  OrderID       Date CustomerID Product  Quantity  UnitPrice ShippingAddress PaymentMethod OrderStatus TrackingNumber  ItemsInCart CouponCode ReferralSource  TotalPrice
ORD200000 2023-01-04     C72649 Monitor         5     570.62     928 Main St    Debit Card     Shipped    TRK37947903            7     SAVE10      Instagram     2853.10
ORD200001 2024-08-23     C75739   Phone         2     151.35     823 Main St        Online     Shipped    TRK91186779            3     SAVE10       Referral      302.70
ORD200002 2024-02-27     C81728  Tablet         5     550.68     512 Main St   Credit Card   Cancelled    TRK42903982            8   FREESHIP          Email     2753.40
ORD200003 2023-10-15     C33540   Chair         1     273.19     275 Main St    Debit Card    Returned    TRK62788070            5     SAVE10       Facebook      273.19
ORD200004 2025-05-08     C81840 Printer         4     626.01     668 Main St        Online   Delivered    TRK29241424            8     SAVE10          Emai

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


The orders table has 14 columns including OrderID, Date, CustomerID,
Product, Quantity, UnitPrice, PaymentMethod, OrderStatus, and TotalPrice.
The dataset contains 1,200 orders spanning from 2023 to 2025.

In [7]:
run_query("""SELECT OrderID, Product, TotalPrice, OrderStatus
FROM orders
LIMIT 5""")

  OrderID Product  TotalPrice OrderStatus
ORD200000 Monitor     2853.10     Shipped
ORD200001   Phone      302.70     Shipped
ORD200002  Tablet     2753.40   Cancelled
ORD200003   Chair      273.19    Returned
ORD200004 Printer     2504.04   Delivered


,OrderID,Product,TotalPrice,OrderStatus
0,ORD200000,Monitor,2853.10,Shipped
1,ORD200001,Phone,302.70,Shipped
2,ORD200002,Tablet,2753.40,Cancelled
3,ORD200003,Chair,273.19,Returned
4,ORD200004,Printer,2504.04,Delivered


Selecting key columns shows a mix of products and statuses across orders.
Not all orders are successful — Cancelled and Returned statuses appear
alongside Shipped and Delivered ones.

 ***Filter with WHERE***

In [ ]:
run_query("""SELECT * FROM orders
WHERE OrderStatus = 'Delivered'""")

  OrderID                Date CustomerID Product  Quantity  UnitPrice ShippingAddress PaymentMethod OrderStatus TrackingNumber  ItemsInCart CouponCode ReferralSource  TotalPrice
ORD200004 2025-05-08 00:00:00     C81840 Printer         4     626.01     668 Main St        Online   Delivered    TRK29241424            8     SAVE10          Email     2504.04
ORD200015 2023-07-17 00:00:00     C39416 Printer         1     473.96     942 Main St          Cash   Delivered    TRK54930938            3       None         Google      473.96
ORD200021 2024-11-17 00:00:00     C99023 Monitor         4     342.95     569 Main St          Cash   Delivered    TRK52646983            6       None         Google     1371.80
ORD200024 2024-10-30 00:00:00     C17470  Laptop         2     230.95     956 Main St        Online   Delivered    TRK28775057            3   WINTER15       Referral      461.90
ORD200025 2024-07-24 00:00:00     C42577 Monitor         1     399.98     709 Main St    Debit Card   Delivere

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200004,2025-05-08 00:00:00,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04
1,ORD200015,2023-07-17 00:00:00,C39416,Printer,1,473.96,942 Main St,Cash,Delivered,TRK54930938,3,None,Google,473.96
2,ORD200021,2024-11-17 00:00:00,C99023,Monitor,4,342.95,569 Main St,Cash,Delivered,TRK52646983,6,None,Google,1371.80
3,ORD200024,2024-10-30 00:00:00,C17470,Laptop,2,230.95,956 Main St,Online,Delivered,TRK28775057,3,WINTER15,Referral,461.90
4,ORD200025,2024-07-24 00:00:00,C42577,Monitor,1,399.98,709 Main St,Debit Card,Delivered,TRK69423327,3,WINTER15,Email,399.98
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
226,ORD201166,2023-07-31 00:00:00,C95568,Chair,3,324.47,971 Main St,Cash,Delivered,TRK37075292,4,WINTER15,Email,973.41
227,ORD201168,2023-04-08 00:00:00,C89559,Laptop,2,331.42,297 Main St,Credit Card,Delivered,TRK82838108,6,WINTER15,Google,662.84
228,ORD201171,2023-08-28 00:00:00,C39911,Phone,4,518.40,699 Main St,Credit Card,Delivered,TRK80431998,7,SAVE10,Facebook,2073.60
229,ORD201187,2024-03-13 00:00:00,C26340,Desk,2,184.60,562 Main St,Gift Card,Delivered,TRK59962703,7,WINTER15,Email,369.20


Only 231 out of 1,200 orders are Delivered — just 19.25% of all orders.
The majority of orders are still Pending, Shipped, Cancelled, or Returned,
which means the business is successfully completing less than 1 in 5 orders.

In [ ]:
run_query("""SELECT OrderID, CustomerID, Product, TotalPrice
FROM orders
WHERE PaymentMethod = 'Credit Card' AND OrderStatus = 'Cancelled'""")

  OrderID CustomerID Product  TotalPrice
ORD200002     C81728  Tablet     2753.40
ORD200083     C15681  Laptop     2161.64
ORD200089     C17181 Printer     2276.65
ORD200103     C62656   Phone      873.72
ORD200106     C73514    Desk     1395.15
ORD200116     C47761 Printer     1165.40
ORD200147     C46316  Laptop      930.36
ORD200148     C35100   Phone      639.82
ORD200159     C48965 Printer      299.35
ORD200161     C35091 Monitor      211.14
ORD200202     C36938 Monitor      761.94
ORD200212     C53215 Monitor      265.38
ORD200228     C57867   Chair     1543.68
ORD200257     C89732   Phone      406.22
ORD200266     C38840   Chair     2332.28
ORD200275     C80735   Chair      130.14
ORD200355     C18353  Tablet     1299.99
ORD200365     C56796   Phone      678.38
ORD200368     C85322 Monitor      813.06
ORD200472     C60029   Chair     1524.36
ORD200527     C27202   Chair     3267.35
ORD200564     C93309    Desk     1821.40
ORD200594     C40140   Phone      363.20
ORD200616     C4

,OrderID,CustomerID,Product,TotalPrice
0,ORD200002,C81728,Tablet,2753.40
1,ORD200083,C15681,Laptop,2161.64
2,ORD200089,C17181,Printer,2276.65
3,ORD200103,C62656,Phone,873.72
4,ORD200106,C73514,Desk,1395.15
5,ORD200116,C47761,Printer,1165.40
6,ORD200147,C46316,Laptop,930.36
7,ORD200148,C35100,Phone,639.82
8,ORD200159,C48965,Printer,299.35
9,ORD200161,C35091,Monitor,211.14


Several Credit Card orders were cancelled. This is financially damaging
because cancelled Credit Card transactions may involve refund processing
costs. The business should investigate the reason behind these cancellations
to reduce revenue loss.

In [ ]:
run_query("""SELECT OrderID, Date, Quantity, UnitPrice, TotalPrice
FROM orders
WHERE Product = 'Laptop'""")

  OrderID                Date  Quantity  UnitPrice  TotalPrice
ORD200006 2025-06-17 00:00:00         1     664.42      664.42
ORD200013 2023-08-30 00:00:00         3     201.49      604.47
ORD200024 2024-10-30 00:00:00         2     230.95      461.90
ORD200026 2024-03-06 00:00:00         5      59.55      297.75
ORD200036 2025-05-08 00:00:00         2     534.43     1068.86
ORD200037 2024-10-14 00:00:00         2      76.74      153.48
ORD200038 2024-12-19 00:00:00         1     245.89      245.89
ORD200076 2024-03-31 00:00:00         4     382.88     1531.52
ORD200083 2023-08-22 00:00:00         4     540.41     2161.64
ORD200091 2023-03-15 00:00:00         5     492.75     2463.75
ORD200102 2023-03-08 00:00:00         2      70.34      140.68
ORD200108 2024-12-28 00:00:00         5      48.38      241.90
ORD200113 2024-01-17 00:00:00         5     190.08      950.40
ORD200122 2024-04-15 00:00:00         2     533.02     1066.04
ORD200129 2023-10-27 00:00:00         4     263.93     

,OrderID,Date,Quantity,UnitPrice,TotalPrice
0,ORD200006,2025-06-17 00:00:00,1,664.42,664.42
1,ORD200013,2023-08-30 00:00:00,3,201.49,604.47
2,ORD200024,2024-10-30 00:00:00,2,230.95,461.90
3,ORD200026,2024-03-06 00:00:00,5,59.55,297.75
4,ORD200036,2025-05-08 00:00:00,2,534.43,1068.86
...,...,...,...,...,...
168,ORD201156,2023-07-19 00:00:00,4,690.78,2763.12
169,ORD201163,2024-06-11 00:00:00,3,306.29,918.87
170,ORD201168,2023-04-08 00:00:00,2,331.42,662.84
171,ORD201179,2023-01-08 00:00:00,5,518.55,2592.75


Laptop unit prices range widely — from as low as $59.55 to $664.42 —
suggesting multiple laptop models at different price points. Some customers
purchase up to 5 units in a single order, indicating possible bulk or
business purchases.

***Sort with ORDER BY***

In [ ]:
run_query("""SELECT OrderID, Product, TotalPrice
FROM orders
ORDER BY TotalPrice DESC
LIMIT 10""")

  OrderID Product  TotalPrice
ORD200789  Tablet     3456.40
ORD201122 Monitor     3390.95
ORD200632  Laptop     3390.80
ORD200469   Chair     3384.90
ORD200328  Tablet     3370.20
ORD200107 Printer     3353.75
ORD200326  Laptop     3352.40
ORD201065 Printer     3334.00
ORD201031   Phone     3322.55
ORD200463  Laptop     3313.90


,OrderID,Product,TotalPrice
0,ORD200789,Tablet,3456.40
1,ORD201122,Monitor,3390.95
2,ORD200632,Laptop,3390.80
3,ORD200469,Chair,3384.90
4,ORD200328,Tablet,3370.20
5,ORD200107,Printer,3353.75
6,ORD200326,Laptop,3352.40
7,ORD201065,Printer,3334.00
8,ORD201031,Phone,3322.55
9,ORD200463,Laptop,3313.90


The highest value order is $3,456.40 for a Tablet. All top 10 orders
exceed $3,300 and include Tablets, Monitors, Laptops, Chairs, and Printers.
These high-value customers should be targeted for loyalty and retention
programs.

In [ ]:
run_query("""SELECT OrderID, Product, TotalPrice
FROM orders
ORDER BY TotalPrice ASC
LIMIT 10""")

  OrderID Product  TotalPrice
ORD201161   Phone       11.39
ORD200863   Phone       14.06
ORD200240  Tablet       17.24
ORD200542  Tablet       17.98
ORD200336  Laptop       18.20
ORD200776  Tablet       21.19
ORD201025   Chair       23.53
ORD200690 Monitor       24.48
ORD200926    Desk       26.95
ORD200473   Chair       29.86


,OrderID,Product,TotalPrice
0,ORD201161,Phone,11.39
1,ORD200863,Phone,14.06
2,ORD200240,Tablet,17.24
3,ORD200542,Tablet,17.98
4,ORD200336,Laptop,18.20
5,ORD200776,Tablet,21.19
6,ORD201025,Chair,23.53
7,ORD200690,Monitor,24.48
8,ORD200926,Desk,26.95
9,ORD200473,Chair,29.86


The lowest value order is just $11.39 for a Phone. The bottom 10 orders
are all under $30, likely representing single-unit purchases of budget
items or heavily discounted products using coupon codes.

In [ ]:
run_query("""SELECT OrderID, Date, Product, TotalPrice
FROM orders
WHERE OrderStatus = 'Delivered'
ORDER BY Date DESC""")

  OrderID                Date Product  TotalPrice
ORD200882 2025-06-28 00:00:00   Phone      548.24
ORD200226 2025-06-27 00:00:00    Desk     1764.15
ORD200918 2025-06-24 00:00:00  Laptop     1841.25
ORD200604 2025-06-22 00:00:00   Chair       97.86
ORD201044 2025-06-22 00:00:00 Printer       38.49
ORD200468 2025-06-21 00:00:00   Phone      418.36
ORD200583 2025-06-15 00:00:00   Chair      463.11
ORD200837 2025-06-14 00:00:00   Chair     3277.75
ORD200854 2025-06-13 00:00:00  Laptop     1117.80
ORD200285 2025-06-10 00:00:00   Phone      338.31
ORD200200 2025-06-07 00:00:00  Tablet     2166.30
ORD200870 2025-06-01 00:00:00   Phone     1880.36
ORD200140 2025-05-31 00:00:00   Chair     1584.42
ORD200261 2025-05-28 00:00:00   Phone      868.25
ORD201025 2025-05-17 00:00:00   Chair       23.53
ORD200508 2025-05-14 00:00:00   Phone      558.27
ORD200883 2025-05-13 00:00:00 Printer     2807.40
ORD201145 2025-05-12 00:00:00    Desk     1714.59
ORD200004 2025-05-08 00:00:00 Printer     2504.04


,OrderID,Date,Product,TotalPrice
0,ORD200882,2025-06-28 00:00:00,Phone,548.24
1,ORD200226,2025-06-27 00:00:00,Desk,1764.15
2,ORD200918,2025-06-24 00:00:00,Laptop,1841.25
3,ORD200604,2025-06-22 00:00:00,Chair,97.86
4,ORD201044,2025-06-22 00:00:00,Printer,38.49
...,...,...,...,...
226,ORD200829,2023-02-05 00:00:00,Monitor,76.93
227,ORD200692,2023-01-23 00:00:00,Desk,984.20
228,ORD201080,2023-01-20 00:00:00,Laptop,2323.70
229,ORD201008,2023-01-13 00:00:00,Monitor,1232.64


The most recent delivered order was on June 28, 2025 — a Phone worth
$548.24. Delivered orders include all product types with no single product
dominating, showing balanced delivery performance across categories.

***Aggregate with COUNT, SUM, AVG***

In [ ]:
run_query("SELECT COUNT(*) AS TotalOrders FROM orders")

 TotalOrders
        1200


,TotalOrders
0,1200


The dataset contains exactly 1,200 orders. This is a solid sample size
for drawing reliable business insights across all product categories,
payment methods, and order statuses.

In [ ]:
run_query("SELECT SUM(TotalPrice) AS TotalRevenue FROM orders")

 TotalRevenue
   1264761.96


,TotalRevenue
0,1264761.96


Total gross revenue across all 1,200 orders is $1,264,761.96. This
includes cancelled and returned orders, so actual collected revenue
is significantly lower than this figure.

In [ ]:
run_query("SELECT AVG(TotalPrice) AS AverageRevenue FROM orders")

 AverageRevenue
      1053.9683


,AverageRevenue
0,1053.9683


The average order value is $1,053.97. This relatively high average is
likely driven by customers purchasing multiple units per order of
electronics and furniture items.

In [ ]:
run_query("""SELECT SUM(TotalPrice) AS DeliveredRevenue
FROM orders
WHERE OrderStatus = 'Delivered'""")

 DeliveredRevenue
        242600.32


,DeliveredRevenue
0,242600.32


Only $242,600.32 out of $1,264,761.96 total revenue comes from Delivered
orders — just 19.2%. Over 80% of gross revenue is tied up in Pending,
Shipped, Cancelled, or Returned orders, which represents a significant
business risk.

***GROUP BY***

In [ ]:
run_query("""SELECT OrderStatus, COUNT(*) AS OrderCount
FROM orders
GROUP BY OrderStatus
ORDER BY OrderCount DESC""")

OrderStatus  OrderCount
  Cancelled         250
   Returned         247
    Pending         237
    Shipped         235
  Delivered         231


,OrderStatus,OrderCount
0,Cancelled,250
1,Returned,247
2,Pending,237
3,Shipped,235
4,Delivered,231


Result:
  Cancelled  250
   Returned  247
    Pending  237
    Shipped  235
  Delivered  231

Cancelled orders are the highest at 250, and Delivered orders are the
lowest at 231. The fact that more orders are cancelled than delivered is
a critical business problem that needs immediate attention.

In [ ]:
run_query("""SELECT Product, SUM(TotalPrice) AS TotalRevenue
FROM orders
GROUP BY Product
ORDER BY TotalRevenue DESC""")

Product  TotalRevenue
  Chair     195620.11
Printer     195612.61
 Laptop     192126.56
 Tablet     186568.95
Monitor     175651.41
   Desk     167459.93
  Phone     151722.39


,Product,TotalRevenue
0,Chair,195620.11
1,Printer,195612.61
2,Laptop,192126.56
3,Tablet,186568.95
4,Monitor,175651.41
5,Desk,167459.93
6,Phone,151722.39


Result:
  Chair    195,620.11
Printer    195,612.61
 Laptop    192,126.56
 Tablet    186,568.95
Monitor    175,651.41
   Desk    167,459.93
  Phone    151,722.39

Chair and Printer are virtually tied at the top with ~$195,000 each.
Revenue is fairly evenly spread across all 7 products — no single product
dominates, showing a well-diversified product portfolio.

In [ ]:
run_query("""
SELECT Product,
       COUNT(*) AS TotalOrders,
       SUM(TotalPrice) AS TotalRevenue,
       AVG(TotalPrice) AS AvgOrderValue
FROM orders
GROUP BY Product
ORDER BY TotalRevenue DESC
""")

Product  TotalOrders  TotalRevenue  AvgOrderValue
  Chair          178     195620.11    1098.989382
Printer          181     195612.61    1080.732652
 Laptop          173     192126.56    1110.558150
 Tablet          179     186568.95    1042.284637
Monitor          163     175651.41    1077.616012
   Desk          170     167459.93     985.058412
  Phone          156     151722.39     972.579423


,Product,TotalOrders,TotalRevenue,AvgOrderValue
0,Chair,178,195620.11,1098.989382
1,Printer,181,195612.61,1080.732652
2,Laptop,173,192126.56,1110.558150
3,Tablet,179,186568.95,1042.284637
4,Monitor,163,175651.41,1077.616012
5,Desk,170,167459.93,985.058412
6,Phone,156,151722.39,972.579423


Result:
Product  TotalOrders  TotalRevenue  AvgOrderValue
  Chair          178    195620.11        1098.99
Printer          181    195612.61        1080.73
 Laptop          173    192126.56        1110.56
 Tablet          179    186568.95        1042.28
Monitor          163    175651.41        1077.62
   Desk          170    167459.93         985.06
  Phone          156    151722.39         972.58

Laptop has the highest average order value at $1,110.56 despite not
having the most orders. Phone has both the fewest orders (156) and the
lowest average value ($972.58), making it the weakest performer overall.
Printer has the most orders at 181 but a lower average value than Laptop.

In [ ]:
run_query("""
SELECT ReferralSource, COUNT(*) AS OrderCount
FROM orders
GROUP BY ReferralSource
ORDER BY OrderCount DESC
""")

ReferralSource  OrderCount
     Instagram         259
         Email         250
        Google         241
      Facebook         228
      Referral         222


,ReferralSource,OrderCount
0,Instagram,259
1,Email,250
2,Google,241
3,Facebook,228
4,Referral,222


Result:
 Instagram  259
     Email  250
    Google  241
  Facebook  228
  Referral  222

Instagram is the top customer acquisition channel with 259 orders.
All 5 channels perform similarly with a narrow gap of only 37 orders
between the highest and lowest, suggesting a balanced multi-channel
marketing strategy.

In [ ]:
run_query("""SELECT PaymentMethod, AVG(TotalPrice) AS AvgOrderValue
FROM orders
GROUP BY PaymentMethod
ORDER BY AvgOrderValue DESC""")

PaymentMethod  AvgOrderValue
  Credit Card    1127.553974
    Gift Card    1070.973565
         Cash    1056.041829
       Online    1017.220698
   Debit Card    1001.556810


,PaymentMethod,AvgOrderValue
0,Credit Card,1127.553974
1,Gift Card,1070.973565
2,Cash,1056.041829
3,Online,1017.220698
4,Debit Card,1001.556810


Result:
Credit Card   1127.55
  Gift Card   1070.97
       Cash   1056.04
     Online   1017.22
 Debit Card   1001.56

Credit Card customers spend the most on average at $1,127.55, while
Debit Card customers spend the least at $1,001.56. The $126 gap suggests
Credit Card users tend to purchase higher-value items or larger quantities.

***HAVING***

In [ ]:
run_query("""SELECT Product, SUM(TotalPrice) AS TotalRevenue
FROM orders
GROUP BY Product
HAVING SUM(TotalPrice) > 50000
ORDER BY TotalRevenue DESC""")

Product  TotalRevenue
  Chair     195620.11
Printer     195612.61
 Laptop     192126.56
 Tablet     186568.95
Monitor     175651.41
   Desk     167459.93
  Phone     151722.39


,Product,TotalRevenue
0,Chair,195620.11
1,Printer,195612.61
2,Laptop,192126.56
3,Tablet,186568.95
4,Monitor,175651.41
5,Desk,167459.93
6,Phone,151722.39


All 7 products exceed $50,000 in total revenue, confirming every product
category is commercially strong. Chair leads at $195,620 and Phone is
lowest at $151,722 — but even the lowest product generates over 3x the
threshold, showing a healthy product lineup.

In [ ]:
run_query("""SELECT PaymentMethod, COUNT(*) AS UsageCount
FROM orders
GROUP BY PaymentMethod
HAVING COUNT(*) > 200""")

PaymentMethod  UsageCount
         Cash         246
  Credit Card         234
   Debit Card         232
    Gift Card         230
       Online         258


,PaymentMethod,UsageCount
0,Cash,246
1,Credit Card,234
2,Debit Card,232
3,Gift Card,230
4,Online,258


Result:
       Cash  246
Credit Card  234
 Debit Card  232
  Gift Card  230
     Online  258

All 5 payment methods are used more than 200 times each. Online is the
most frequently used at 258 transactions and Gift Card the least at 230.
The near-equal usage across all methods shows customers have no strong
preference — the business should continue supporting all payment options.